exporter/clickhouselogsexporter/promoted.go

```diff
@@ -0,0 +1,116 @@
+)
+
+// buildPromotedAndPruneBody extracts promoted paths from body and returns them as a separate map.
+// It mutates the body in place to remove extracted entries.
+func buildPromotedAndPruneBody(body pcommon.Value, promotedPaths map[string]struct{}) pcommon.Value {
+	promoted := pcommon.NewValueMap()
+	if body.Type() != pcommon.ValueTypeMap || len(promotedPaths) == 0 {
+		return promoted
+	}
+
+	bm := body.Map()
+	pm := promoted.Map()
+	for path := range promotedPaths {
+		// For each path, extract with literal preference at every level
+		handleSinglePath(bm, pm, path, path)
+	}
+
+	// Clean up empty maps that were created during extraction
+	cleanupEmptyMaps(bm)
+
+	return promoted
+}
+
+// handleSinglePath walks the map according to remainingPath and extracts the value into promotedMap at fullPath.
+// Literal preference: at the current map level, if a literal key equal to remainingPath exists, use it and stop descending.
+func handleSinglePath(bodyMap pcommon.Map, promotedMap pcommon.Map, fullPath string, remainingPath string) {
+	// Prefer literal match of the entire remaining path at this level
+	if v, ok := bodyMap.Get(remainingPath); ok {
+		dst := promotedMap.PutEmpty(fullPath)
+		v.CopyTo(dst)
+		bodyMap.Remove(remainingPath)
+		return
+	}
+
+	// If no dot remains, try direct segment
+	dot := strings.IndexByte(remainingPath, '.')
+	if dot == -1 {
+		if v, ok := bodyMap.Get(remainingPath); ok {
+			dst := promotedMap.PutEmpty(fullPath)
+			v.CopyTo(dst)
+			bodyMap.Remove(remainingPath)
+		}
+		return
+	}
+
+	head := remainingPath[:dot]
+	tail := remainingPath[dot+1:]
+
+	// If there is no map under head, but there are dotted literal siblings like head.xxx,
+	// materialize a submap under head and move those keys into it for consistent traversal.
+	if v, ok := bodyMap.Get(head); !ok || v.Type() != pcommon.ValueTypeMap {
+		// scan for dotted literals
+		var foundPrefixed bool
+		bodyMap.Range(func(k string, vv pcommon.Value) bool {
+			if strings.HasPrefix(k, head+".") {
+				foundPrefixed = true
+				return false
+			}
+			return true
+		})
+		if foundPrefixed {
+			// create submap and move entries
+			sub := bodyMap.PutEmptyMap(head)
+			var keysToMove []string
+			bodyMap.Range(func(k string, vv pcommon.Value) bool {
+				if strings.HasPrefix(k, head+".") {
+					keysToMove = append(keysToMove, k)
+				}
+				return true
+			})
+			for _, k := range keysToMove {
+				vv, _ := bodyMap.Get(k)
+				// move only one level: head.rest -> sub[rest]
+				rest := k[len(head)+1:]
+				dst := sub.PutEmpty(rest)
+				vv.CopyTo(dst)
+				bodyMap.Remove(k)
+			}
+		}
+	}
+
+	if v, ok := bodyMap.Get(head); ok && v.Type() == pcommon.ValueTypeMap {
+		// Recurse into nested map
+		handleSinglePath(v.Map(), promotedMap, fullPath, tail)
+	}
+}
+
+// cleanupEmptyMaps removes empty maps from the body
+func cleanupEmptyMaps(bodyMap pcommon.Map) {
+	// First, recursively clean up nested maps
+	bodyMap.Range(func(k string, v pcommon.Value) bool {
```

Comment:
what is the exact usecase of traversing the entire json again, this seems like something very expensive as it will traverse all big json.

Reply:
Yes, planning to merge them together